### ISSUE TITLE
schema publish error: read ECONNRESET

### ISSUE THREAD

**Issue by mydearxym** [apollographql/apollo-cli#523]
hi there, i am using the new apollo-cli to upload schema  but aways get `ECCONNRESET` error: 

![image](https://user-images.githubusercontent.com/6184465/43506524-ec0d8168-959c-11e8-8527-873ce20900ef.png)

**Comment by mydearxym** [comment]
Nope,  i am using VPN all the time, it's not the firewall issue

**Comment by mydearxym** [comment]
it maybe the proxy issue caused by china firewall, the **apollo-engine stack seems not validable at china** ...

use VPN: 
```text
Error: request to https://engine-graphql.apollographql.com/api/graphql/ failed, reason: read ECONNRESET
```
without VPN:
```text
Error: request to https://engine-graphql.apollographql.com/api/graphql/ failed, reason: connect ETIMEDOUT 130.211.22.123:443
```
Any chance i can push to ` http://engine-graphql...` instead of `https://engine-graphql.` ?

**Comment by zionts** [comment]
I think this may be an issue with the Apollo CLI not supporting corporate proxies? Node fetch doesn't support this by default: https://github.com/TooTallNate/node-https-proxy-agent

**Comment by mydearxym** [comment]
@zionts  i move the publish logic to travis, and problem solved :)

**Comment by LPink777** [comment]
I have the same problem when i start fullstack-tutorial and push service
![image](https://user-images.githubusercontent.com/19523872/50539098-80d5eb80-0bb5-11e9-8da2-449021f3fdc3.png)
what should i do?@zionts@mydearxym

**Comment by FX-HAO** [comment]
This issue should not be closed, we need something like `--proxy` flag or `https_proxy` env.

**Comment by sirkay777** [comment]
### Workaround

**Step 1**
 `npm i -g proxy-agent`

**Step 2**
append code below to `/usr/local/lib/node_modules/apollo/node_modules/apollo-env/lib/fetch/fetch.js
` (this path is from my Mac, change global node_modules according to your system)
```javascript
var ProxyAgent = require("proxy-agent");
var proxy = "socks://127.0.0.1:1080";  // or "http://127.0.0.1:8001"
var agent = new ProxyAgent(proxy);
exports.fetch = function(url, options) {
  return require("node-fetch").default(url, { ...options, agent });
};
```
**Why**
command `apollo` depends on `apollo-env` which use `node-fetch`
`node-fectch` can accept a agent option for proxy : https://www.npmjs.com/package/node-fetch#options
`proxy-agent` provides such agent (http,https,socks5,etc...) : https://github.com/TooTallNate/node-proxy-agent

**Comment by yangchongduo** [comment]
@LPink777  How to deal with the same problem

**Comment by yangchongduo** [comment]
@sirkay777 request to http://localhost:4000/ failed, reason: connect ECONNREFUSED 127.0.0.1:1080

**Comment by yangchongduo** [comment]
var proxy = "socks://127.0.0.1:1080";  // or "http://127.0.0.1:8001"

**Comment by sirkay777** [comment]
@yangchongduo 

> connect ECONNREFUSED 127.0.0.1:1080

Make sure you have a socks proxy client running on port 1080 (shadowsocks, v2ray etc) 

> request to http://localhost:4000/ failed

Be aware the above modification to `fetch.js` may cause unexpected  behavior. Apollo tool depend on this to make http request.  Now all request will go through the socks proxy. Even `request to http://localhost:4000/`. If your proxy client cannot handle this, you should proxy only the request to `https://engine-graphql.apollographql.com'`:
```javascript
exports.fetch = function(url, options) {
  if(url.startsWith('https://engine-graphql.apollographql.com')){
  	return require("node-fetch").default(url, { ...options, agent });
  }
  return require("node-fetch").default;
};
```

**Comment by wZi** [comment]
@sirkay777 
you save me

**Comment by jiayechao** [comment]
> @yangchongduo
> 
> > connect ECONNREFUSED 127.0.0.1:1080
> 
> Make sure you have a socks proxy client running on port 1080 (shadowsocks, v2ray etc)
> 
> > request to http://localhost:4000/ failed
> 
> Be aware the above modification to `fetch.js` may cause unexpected behavior. Apollo tool depend on this to make http request. Now all request will go through the socks proxy. Even `request to http://localhost:4000/`. If your proxy client cannot handle this, you should proxy only the request to `https://engine-graphql.apollographql.com'`:
> 
> ```js
> exports.fetch = function(url, options) {
>   if(url.startsWith('https://engine-graphql.apollographql.com')){
>   	return require("node-fetch").default(url, { ...options, agent });
>   }
>   return require("node-fetch").default;
> };
> ```

大佬，我用的是1.6版本的apollo，并没有找到`/usr/local/lib/node_modules/apollo/node_modules/apollo-env/lib/fetch/fetch.js`这个文件

**Comment by xw616525957** [comment]
@sirkay777 ,Thanks,I update it as follow
### set terminal proxy  for mac
```bash
alias proxy='export all_proxy=socks5://127.0.0.1:1080'
alias unproxy='unset all_proxy'
```
when want to proxy:
```bash
proxy
```
when want unuse proxy
```bash
unproxy
```
I don't know why this does not work.  For me ,below is work!
### Step 2
append code below to /usr/local/lib/node_modules/apollo/node_modules/apollo-env/lib/fetch/fetch.js (this path is from my Mac, change global node_modules according to your system)
```js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ProxyAgent = require("proxy-agent");
var proxy = "socks://127.0.0.1:1080";  // or "http://127.0.0.1:8001"
var agent = new ProxyAgent(proxy);
var node_fetch_1 = require("node-fetch");
//exports.fetch = node_fetch_1.default;
exports.fetch = function(url, options) {
	if(url.indexOf('localhost')>-1){
		return node_fetch_1.default(url,options);
	}
  return node_fetch_1.default(url, { ...options, agent });
};
exports.Request = node_fetch_1.Request;
exports.Response = node_fetch_1.Response;
exports.Headers = node_fetch_1.Headers;
//# sourceMappingURL=fetch.js.map
```

### For apollo-vscode
~/.vscode/extensions/apollographql.vscode-apollo-1.4.0/node_modules/apollo-env/lib/fetch/fetch.js
```js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });

var ProxyAgent = require("/usr/local/lib/node_modules/proxy-agent"); //use global proxy-agent
var proxy = "socks://127.0.0.1:1080";  // or "http://127.0.0.1:8001"
var agent = new ProxyAgent(proxy);

var node_fetch_1 = require("node-fetch");
//exports.fetch = node_fetch_1.default;
exports.fetch = function(url, options) {
	if(url.indexOf('localhost')>-1){
		return node_fetch_1.default(url,options);
	}
  return node_fetch_1.default(url, { ...options, agent });
};
exports.Request = node_fetch_1.Request;
exports.Response = node_fetch_1.Response;
exports.Headers = node_fetch_1.Headers;
//# sourceMappingURL=fetch.js.map
```

**Comment by steveoOn** [comment]
> @sirkay777 ,Thanks,I update it as follow
> 
> ### set terminal proxy for mac
> ```shell
> alias proxy='export all_proxy=socks5://127.0.0.1:1080'
> alias unproxy='unset all_proxy'
> ```
> 
> when want to proxy:
> 
> ```shell
> proxy
> ```
> 
> when want unuse proxy
> 
> ```shell
> unproxy
> ```
> 
> I don't know why this does not work. For me ,below is work!
> 
> ### Step 2
> append code below to /usr/local/lib/node_modules/apollo/node_modules/apollo-env/lib/fetch/fetch.js (this path is from my Mac, change global node_modules according to your system)
> 
> ```js
> "use strict";
> Object.defineProperty(exports, "__esModule", { value: true });
> var ProxyAgent = require("proxy-agent");
> var proxy = "socks://127.0.0.1:1080";  // or "http://127.0.0.1:8001"
> var agent = new ProxyAgent(proxy);
> var node_fetch_1 = require("node-fetch");
> //exports.fetch = node_fetch_1.default;
> exports.fetch = function(url, options) {
> 	if(url.indexOf('localhost')>-1){
> 		return node_fetch_1.default(url,options);
> 	}
>   return node_fetch_1.default(url, { ...options, agent });
> };
> exports.Request = node_fetch_1.Request;
> exports.Response = node_fetch_1.Response;
> exports.Headers = node_fetch_1.Headers;
> //# sourceMappingURL=fetch.js.map
> ```
> 
> ### For apollo-vscode
> ~/.vscode/extensions/apollographql.vscode-apollo-1.4.0/node_modules/apollo-env/lib/fetch/fetch.js
> 
> ```js
> "use strict";
> Object.defineProperty(exports, "__esModule", { value: true });
> 
> var ProxyAgent = require("/usr/local/lib/node_modules/proxy-agent"); //use global proxy-agent
> var proxy = "socks://127.0.0.1:1080";  // or "http://127.0.0.1:8001"
> var agent = new ProxyAgent(proxy);
> 
> var node_fetch_1 = require("node-fetch");
> //exports.fetch = node_fetch_1.default;
> exports.fetch = function(url, options) {
> 	if(url.indexOf('localhost')>-1){
> 		return node_fetch_1.default(url,options);
> 	}
>   return node_fetch_1.default(url, { ...options, agent });
> };
> exports.Request = node_fetch_1.Request;
> exports.Response = node_fetch_1.Response;
> exports.Headers = node_fetch_1.Headers;
> //# sourceMappingURL=fetch.js.map
> ```

it's working for me, thanks~

**Comment by lastingman** [comment]
This issue should not be closed, I need to implement the solution everytime I change my version of node. It is quite annoying.

**Comment by BobLamarley** [comment]
Same here with this config behind a f****g corporate proxy ^^
![image](https://user-images.githubusercontent.com/7857364/72975812-bd344500-3dd1-11ea-9012-8b92ba43c146.png)
Here is what i got and it's quite annoying, i can't publish my schema to apollo graph manager :(
![image](https://user-images.githubusercontent.com/7857364/72975907-df2dc780-3dd1-11ea-856f-f3a6b4fa3fe6.png)

**Comment by kuangbeibei** [comment]
Follow the steps from @sirkay777 
![image](https://user-images.githubusercontent.com/10113668/84742045-8100bb80-afe2-11ea-9354-75ea2c513e46.png)

Thanks

**Comment by tsareg** [comment]
Why this is closed? Modifying node_modules is definitely not an answer!

**Comment by antonmo46** [comment]
another work around for your corporate proxy

```
  script:
    - npm install -D apollo
    - npm install -D proxy-agent
    - printf "%s\n" >> node_modules/apollo-env/lib/fetch/fetch.js
    - printf "var ProxyAgent=require('proxy-agent');var agent=new ProxyAgent('http://my-corporate-proxy.com:3128/');exports.fetch=function(url,options){return require('node-fetch').default(url,{...options,agent})};" >> node_modules/apollo-env/lib/fetch/fetch.js
    - npx apollo service:check --localSchemaFile=$(find src/main/resources/graphql -name "*.graphqls" | tr "\n" "," | sed 's/.$//')
```

**Comment by glasser** [comment]
Hi, I believe I've fixed this issue. Since I don't work in a corporate proxy environment myself, I'd appreciate help testing the alpha from folks who've faced this issue.  You can install the alpha with `npm i apollo@2.32.0-alpha.0`.  This uses the `global-agent` npm package with the standard env vars `HTTPS_PROXY` and `NO_PROXY` (no `GLOBAL_AGENT_` prefix needed). Will publish this as non-alpha once I get confirmation it works for folks!

**Comment by glasser** [comment]
Fixed (hopefully!) in `apollo@0.32.0`.
